spec.md

```diff
@@ -0,0 +1,954 @@
+
+#### Timeouts
+
+Any of the RPCs defined in this spec MAY timeout and MAY be retried.
+The CMI-Client may choose the maximum time it is willing to wait for a call, how long it waits between retries, and how many time it retries (these values are not negotiated between plugin and CO).
+
+Idempotency requirements ensure that a retried call with the same fields continues where it left off when retried.
+
+### Error Scheme
+
+All CMI API calls defined in this spec MUST return a [standard gRPC status](https://github.com/grpc/grpc/blob/master/src/proto/grpc/status/status.proto).
+Most gRPC libraries provide helper methods to set and read the status fields.
+
+The status `code` MUST contain a [canonical error code](https://github.com/grpc/grpc-go/blob/master/codes/codes.go). CMI-Clients MUST handle all valid error codes. Each RPC defines a set of gRPC error codes that MUST be returned by the plugin when specified conditions are encountered. In addition to those, if the conditions defined below are encountered, the plugin MUST return the associated gRPC error code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Missing required field | 3 INVALID_ARGUMENT | Indicates that a required field is missing from the request. More human-readable information MAY be provided in the `status.message` field. | Caller MUST fix the request by adding the missing required field before retrying. |
+| Invalid or unsupported field in the request | 3 INVALID_ARGUMENT | Indicates that the one or more fields in this field is either not allowed by the Plugin or has an invalid value. More human-readable information MAY be provided in the gRPC `status.message` field. | Caller MUST fix the field before retrying. |
+| Permission denied | 7 PERMISSION_DENIED | The Plugin is able to derive or otherwise infer an identity from the secrets present within an RPC, but that identity does not have permission to invoke the RPC. | System administrator SHOULD ensure that requisite permissions are granted, after which point the caller MAY retry the attempted RPC. |
+| Operation pending for machine | 10 ABORTED | Indicates that there is already an operation pending for the specified machine. In general the CMI-Client is responsible for ensuring that there is no more than one call "in-flight" per machine at a given time. However, in some circumstances, the CMI-Client MAY lose state (for example when the CMI-Client crashes and restarts), and MAY issue multiple calls simultaneously for the same machine. The Plugin, SHOULD handle this as gracefully as possible, and MAY return this error code to reject secondary calls. | Caller SHOULD ensure that there are no other calls pending for the specified machine, and then retry with exponential back off. |
```

Comment:
I was wondering if we could enforce exponential backoff in MCM for errored retries of machine objects. This would avoid queuing machines too often. Especially in cases like machines failing to join due to network issues, causing it to recreate several machines.

Reply:
In that case, shall we split the sentence and keep the SHOULD only for the ensuring no other calls are pending part and say that it is *recommended* that the retry should be governed by an exponential backoff?

Reply:
Yes sure. Makes sense. Will create a commit with all your suggested changes soon.

Reply:
I also think it's important to keep the handling of `ABORTED` as this situation can actually occur, and even explicitly describe it in the tables listing the errors of different calls, see below. Whether the backoff is exponential or not is a detail and should be omitted IMO. Ideally timeout, number of retries, or even retry strategy (exponential or not) should be configurable on MCM side.

Reply:
@stoyanr - Apologies for the delay in response. I think your idea of having configurable retry strategies on the MCM makes sense. Will Remove the details about retries in this section. Will try to describe the retry strategy in another section. 

WDYT @amshuman-kr , @stoyanr ?

Reply:
Have made the changes - https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR225. And have added a section on retry - https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR205-R212.